opendevin/condenser/condenser.py

```diff
@@ -0,0 +1,147 @@
+from opendevin.core.exceptions import (
+    SummarizeError,
+)
+from opendevin.core.logger import opendevin_logger as logger
+from opendevin.events.action import (
+    AgentSummarizeAction,
+)
+from opendevin.llm.messages import Message
+
+from .prompts import (
+    MESSAGE_SUMMARY_WARNING_FRAC,
+    SUMMARY_PROMPT_SYSTEM,
+    parse_summary_response,
+)
+
+
+class CondenserMixin:
+    """Condenses a group of condensable messages as done by MemGPT."""
+
+    def condense(
+        self,
+        messages: list[Message],
+    ):
+        # Start past the system message, and example messages.,
+        # and collect messages for summarization until we reach the desired truncation token fraction (eg 50%)
+        # Do not allow truncation  for in-context examples of function calling
+        token_counts = [
+            self.get_token_count([message])  # type: ignore
+            for message in messages
+            if message.condensable
+        ]
+        message_buffer_token_count = sum(token_counts)  # no system and example message
+
+        desired_token_count_to_summarize = int(
+            message_buffer_token_count * self.config.message_summary_trunc_tokens_frac  # type: ignore
+        )
+
+        candidate_messages_to_summarize = []
+        tokens_so_far = 0
+        for message in messages:
```

Comment:
We gave up using State, but then we still need to keep untouched the user message that is likely to be a task. That is computed here: https://github.com/OpenDevin/OpenDevin/blob/84a6e90dc2e8b1e096af33f7545fa1969853c7d4/opendevin/controller/state/state.py#L169

Maybe we can set it as non-condensable in the agent, before we get here? Otherwise we're losing relevant information, I think.

Reply:
Yes, something needs to be done about user messages. It is currently losing relevant information. But only marking user messages as non-condensable is not sufficient ig? Some user messages might not contribute much to the relevant information about the final task.?

Currently `summarize_messages()` has `summarized_actions` & `summarized_observations` as args. I was thinking, maybe adding one more arg, `summarized_user_messages` might also help. This will summarize all the user messages till now.

Reply:
>  It is currently losing relevant information. But only marking user messages as non-condensable is not sufficient ig? Some user messages might not contribute much to the relevant information about the final task.?

I suggest to not do it for all user messages, but the most likely task. The method linked above, `get_current_user_intent()`, returns only one user message: what is likely to be the current task. If we set that one as non-condensable, it will already improve the behavior. For the SWE-bench, that method will work to return the initial user message, which is definitely the task itself.

For interactive user sessions, you're right it's more complicated. The method *tries* to guess what's the best user message, and returns the last one after a FinishAction, so it's the user message just after a task ended, and a new one started. That's a best-guess, of course, that the new task is in that message, but IMHO it will do for now. Do you think it's likely it won't work?

> I was thinking, maybe adding one more arg, summarized_user_messages might also help. This will summarize all the user messages till now.

In the previous version (PR 2021), we were summarizing the user messages in a summary of their own, as you suggest, when there were no more agent messages, so at a later stage, but again except the `current user intent`. The current intent as detected (or attempted 😅) remained uncondensed as long as possible. Doing something like this might still be a good option, but I don't think it's necessary for this PR.

The important part, it seems to me, is that if we *miss* the initial user message, then we risk that the agent will not have enough information about the task to perform correctly.